src/ucp/core/ucp_rkey.c

```diff
@@ -213,61 +218,377 @@ UCS_PROFILE_FUNC(ssize_t, ucp_rkey_pack_memh,
+        status      = uct_md_mkey_pack_v2(tl_mds[md_index].md,
+                                          uct_memhs[md_index], &params,
+                                          tl_mkey_buf);
+        if (status != UCS_OK) {
+            result = status;
+            goto out;
+        }
+
+        global_id_size = ucp_memh_global_id_packed_size(md_attr);
+
+        *ucs_serialize_next(&p, uint8_t) = global_id_size;
+
+        global_id_buf = ucs_serialize_next_raw(&p, void, global_id_size);
+        memcpy(global_id_buf, md_attr->global_id, global_id_size);
+
+        ucs_trace("exported mkey[%d]=%s for md[%d]=%s", uct_memh_index,
+                  ucs_str_dump_hex(p, tl_mkey_size, buf, sizeof(buf),
+                                   SIZE_MAX),
+                  md_index, tl_mds[md_index].rsc.md_name);
+        ++uct_memh_index;
+    }
+
+    result = UCS_PTR_BYTE_DIFF(buffer, p);
+
+out:
+    ucs_log_indent(-1);
+    return result;
+}
+
+static size_t
+ucp_memh_packed_size(ucp_mem_h memh, uint64_t flags, int rkey_compat)
+{
+    ucp_context_h context = memh->context;
+
+    if (flags & UCP_MEMH_PACK_FLAG_EXPORT) {
+        ucs_assert(!rkey_compat);
+        return ucp_memh_exported_packed_size(
+                context, context->export_md_map[memh->mem_type]);
+    }
+
+    if (rkey_compat) {
+        return ucp_rkey_packed_size(context, memh->md_map,
+                                    UCS_SYS_DEVICE_ID_UNKNOWN, 0);
+    }
+
+    ucs_fatal("packing rkey using ucp_memh_pack() is unsupported");
+}
+
+static ssize_t ucp_memh_do_pack(ucp_mem_h memh, uint64_t flags,
+                                int rkey_compat, void *memh_buffer)
+{
+    ucp_memory_info_t mem_info;
+
+    if (flags & UCP_MEMH_PACK_FLAG_EXPORT) {
+        return ucp_memh_exported_pack(memh, memh_buffer);
+    }
+
+    if (rkey_compat) {
+        mem_info.type    = memh->mem_type;
+        mem_info.sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
+        return ucp_rkey_pack_memh(memh->context, memh->md_map, memh, &mem_info,
+                                  0, NULL, memh_buffer);
+    }
+
+    ucs_fatal("packing rkey using ucp_memh_pack() is unsupported");
+}
+
 static ucs_status_t
 ucp_memh_pack_internal(ucp_mem_h memh, const ucp_memh_pack_params_t *params,
                        int rkey_compat, void **buffer_p, size_t *buffer_size_p)
 {
     ucp_context_h context = memh->context;
-    ucp_memory_info_t mem_info;
     ucs_status_t status;
     ssize_t packed_size;
-    void *rkey_buffer;
+    void *memh_buffer;
     size_t size;
+    uint64_t flags;
 
-    ucs_trace("packing rkeys for buffer %p memh %p md_map 0x%"PRIx64,
-              ucp_memh_address(memh), memh, memh->md_map);
-
-    if (!rkey_compat) {
-        return UCS_ERR_UNSUPPORTED;
-    }
+    ucs_trace("packing memh %p for buffer %p md_map 0x%" PRIx64
+              " export_md_map 0x%" PRIx64, memh, ucp_memh_address(memh),
+              memh->md_map, context->export_md_map[memh->mem_type]);
```

Comment:
type of the memh is worth mentioning here (external or not)

Reply:
done